EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs

```diff
@@ -126,38 +127,109 @@ private List<ProteinGroup> RunProteinParsimonyEngine()
                             {
-                                peptideInProteinInfo.Add(new Tuple<Protein, DigestionParams, int, int, int, int>(peptide.Peptide.Protein, peptide.Peptide.DigestionParams,
-                                    peptide.Peptide.OneBasedStartResidueInProtein, peptide.Peptide.OneBasedEndResidueInProtein, peptide.Peptide.MissedCleavages, peptide.Notch));
-                            }
+                                var baseSequence = sequenceWithPsmsList[i];
 
-                            // add the protein associations to the PSM
-                            foreach (PeptideSpectralMatch psm in baseSequence.Value)
-                            {
-                                foreach (var proteinInfo in peptideInProteinInfo)
+                                var peptidesWithNotchInfo = baseSequence.Value.SelectMany(p => p.BestMatchingPeptides).Distinct().ToList();
+
+                                // if the base seq has >1 PeptideWithSetMods object and has >0 mods, it might need to be matched to new proteins
+                                if (peptidesWithNotchInfo.Count > 1 && peptidesWithNotchInfo.Any(p => p.Peptide.NumMods > 0))
                                 {
-                                    var originalPep = psm.BestMatchingPeptides.First().Peptide;
-                                    var pep = new PeptideWithSetModifications(proteinInfo.Item1, proteinInfo.Item2, proteinInfo.Item3, proteinInfo.Item4,
-                                        originalPep.CleavageSpecificityForFdrCategory, originalPep.PeptideDescription, proteinInfo.Item5, originalPep.AllModsOneIsNterminus,
-                                        originalPep.NumFixedMods);
-                                    _fdrFilteredPeptides.Add(pep);
-                                    psm.AddProteinMatch((proteinInfo.Item6, pep));
+                                    bool needToAddPeptideToProteinAssociations = false;
+
+                                    // numProteinsForThisBaseSequence is the total number of proteins that this base sequence is a digestion product of
+                                    int numProteinsForThisBaseSequence = peptidesWithNotchInfo.Select(p => p.Peptide.Protein).Distinct().Count();
+
+                                    if (numProteinsForThisBaseSequence == 1)
+                                    {
+                                        continue;
+                                    }
+
+                                    foreach (var psm in baseSequence.Value)
+                                    {
+                                        // numProteinsForThisPsm is the number of proteins that this PSM's peptides are associated with
+                                        int numProteinsForThisPsm = psm.BestMatchingPeptides.Select(p => p.Peptide.Protein).Distinct().Count();
+
+                                        if (numProteinsForThisPsm != numProteinsForThisBaseSequence)
```

Comment:
These two numbers should generally be the same because they are from the same identifications. In what situation they will be different?

Reply:
it's pretty rare. if two proteins have a peptide digest product with the same base sequence (a shared peptide), but only one of them is modified (a non-shared modification), and the modified version is observed in the data, this can affect parsimony. the user can disable whether to ignore mods or not for parsimony